File: app/src/main/java/ru/nsu/bobrofon/easysshfs/EasySSHFSActivity.java
package ru.nsu.bobrofon.easysshfs;

import android.content.Context;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.app.ActionBar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.support.v4.widget.DrawerLayout;
import android.widget.Toast;

import ru.nsu.bobrofon.easysshfs.log.LogFragment;
import ru.nsu.bobrofon.easysshfs.mountpointList.mountpoint.EditFragment;
import ru.nsu.bobrofon.easysshfs.mountpointList.MountpointFragment;


public class EasySSHFSActivity extends ActionBarActivity
	implements NavigationDrawerFragment.NavigationDrawerCallbacks, MountpointFragment.OnFragmentInteractionListener {

	/**
	 * Fragment managing the behaviors, interactions and presentation of the navigation drawer.
	 */
	private NavigationDrawerFragment mNavigationDrawerFragment;

	/**
	 * Used to store the last screen title. For use in {@link #restoreActionBar()}.
	 */
	private CharSequence mTitle;
	private Fragment[] mFragments;
	private static Context mContext;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		new VersionUpdater(getApplicationContext()).update();

		mFragments = new Fragment[] {
			new MountpointFragment(),
			new LogFragment()
		};

		setContentView(R.layout.activity_easy_sshfs);

		mNavigationDrawerFragment = (NavigationDrawerFragment)
			getSupportFragmentManager().findFragmentById(R.id.navigation_drawer);
		mTitle = getTitle();

		// Set up the drawer.
		mNavigationDrawerFragment.setUp(
			R.id.navigation_drawer,
			(DrawerLayout) findViewById(R.id.drawer_layout));

		((MountpointFragment)mFragments[0]).setDrawerStatus(mNavigationDrawerFragment);
		((LogFragment)mFragments[1]).setDrawerStatus(mNavigationDrawerFragment);
	}

	@Override
	public void onResume() {
		super.onResume();
		mContext = getApplicationContext();
	}

	@Override
	public void onPause() {
		super.onPause();
		mContext = null;
	}

	@Override
	public void onNavigationDrawerItemSelected(int position) {
		// update the main content by replacing fragments
		try {
			FragmentManager fragmentManager = getSupportFragmentManager();
			int backStackCount = fragmentManager.getBackStackEntryCount();
			while (backStackCount-- > 0) {
				fragmentManager.popBackStack();
			}

			fragmentManager.beginTransaction()
				.replace(R.id.container, mFragments[position])
				.commit();
		}
		catch (ArrayIndexOutOfBoundsException e) {
			finish();
		}
	}

	public void onSectionAttached(int titleId) {
		mTitle = getString(titleId);
	}

	public void restoreActionBar() {
		ActionBar actionBar = getSupportActionBar();
		actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_STANDARD);
		actionBar.setDisplayShowTitleEnabled(true);
		actionBar.setTitle(mTitle);
	}


	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		if (!mNavigationDrawerFragment.isDrawerOpen()) {
			// Only show items in the action bar relevant to this screen
			// if the drawer is not showing. Otherwise, let the drawer
			// decide what to show in the action bar.
			restoreActionBar();
			return true;
		}
		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		// Handle action bar item clicks here. The action bar will
		// automatically handle clicks on the Home/Up button, so long
		// as you specify a parent activity in AndroidManifest.xml.
		int id = item.getItemId();

		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onFragmentInteraction(int id) {
		EditFragment editFragment = EditFragment.newInstance(id);
		editFragment.setDrawerStatus(mNavigationDrawerFragment);

		FragmentManager fragmentManager = getSupportFragmentManager();
		fragmentManager.beginTransaction()
			.replace(R.id.container, editFragment)
			.addToBackStack(null)
			.commit();
	}

	public static void showToast(final CharSequence message) {
		if (mContext != null) {
			Toast toast = Toast.makeText(mContext, message, Toast.LENGTH_SHORT);
			toast.show();
		}
	}

}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/EasySSHFSActivity.kt
// SPDX-License-Identifier: MIT
package ru.nsu.bobrofon.easysshfs

import android.Manifest
import android.app.UiModeManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.google.android.material.navigation.NavigationView
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import ru.nsu.bobrofon.easysshfs.log.AppLog
import ru.nsu.bobrofon.easysshfs.mountpointlist.MountPointsList
import ru.nsu.bobrofon.easysshfs.settings.SettingsRepository
import ru.nsu.bobrofon.easysshfs.settings.ThemeMode
import ru.nsu.bobrofon.easysshfs.settings.settingsDataStore


private const val TAG = "EasySSHFSActivity"
private const val PERMISSION_REQUEST_CODE = 1

class EasySSHFSActivity : AppCompatActivity() {

    val shell: Shell by lazy { initNewShell() }

    private lateinit var viewModel: EasySSHFSViewModel

    private val drawerLayout: DrawerLayout by lazy { findViewById(R.id.drawer_layout) }
    private val navController: NavController by lazy {
        (supportFragmentManager.findFragmentById(R.id.main_content) as NavHostFragment).navController
    }
    private val navigationView: NavigationView by lazy { findViewById(R.id.nav_view) }

    private lateinit var appBarConfiguration: AppBarConfiguration

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val settingsRepository = SettingsRepository(applicationContext.settingsDataStore)

        viewModel = ViewModelProvider(
            this,
            EasySSHFSViewModel.Factory(
                settingsRepository,
                MountPointsList.instance(applicationContext)
            )
        )[EasySSHFSViewModel::class.java]

        AppLog.instance().addMessage(
            "EasySSHFS ${BuildConfig.VERSION_NAME} " + "(${BuildConfig.VERSION_CODE}) ${BuildConfig.BUILD_TYPE}"
        )
        VersionUpdater(applicationContext).update()

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            // night mode is not persisted by framework, so we need to update it
            // BEFORE content view is set
            val themeMode = runBlocking { settingsRepository.themeMode.first() }
            updateThemeMode(themeMode)
        }

        setContentView(R.layout.activity_easy_sshfs)

        appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.mountpointFragment,
                R.id.logFragment,
                R.id.settingsFragment
            ),
            drawerLayout
        )
        setupActionBarWithNavController(navController, appBarConfiguration)
        navigationView.setupWithNavController(navController)

        val exitMenuItem = navigationView.menu.findItem(R.id.exitAction)
        exitMenuItem.setOnMenuItemClickListener { _ ->
            finish()
            true
        }

        ensureAllPermissionsGranted()

        viewModel.autoMountServiceRequired.observe(this) {
            onAutoMountChanged(it)
        }

        viewModel.themeMode.observe(this) {
            updateThemeMode(it)
        }
    }

    private fun updateThemeMode(mode: ThemeMode) {
        Log.i(TAG, "change app theme mode to '$mode'")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val uiModeManager = getSystemService(UI_MODE_SERVICE) as UiModeManager
            uiModeManager.setApplicationNightMode(mode.nightMode)
        } else {
            AppCompatDelegate.setDefaultNightMode(mode.compatNightMode)
        }
    }

    private fun ensureAllPermissionsGranted() {
        val allPermissions = arrayListOf(
            Manifest.permission.INTERNET,
            Manifest.permission.ACCESS_WIFI_STATE,
            Manifest.permission.ACCESS_NETWORK_STATE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.RECEIVE_BOOT_COMPLETED,
        )
        allPermissions.add(Manifest.permission.READ_EXTERNAL_STORAGE)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            allPermissions.add(Manifest.permission.FOREGROUND_SERVICE)
        }

        val permissions = allPermissions.filter {
            ContextCompat.checkSelfPermission(
                applicationContext, it
            ) != PackageManager.PERMISSION_GRANTED
        }

        if (permissions.isEmpty()) {
            Log.d(TAG, "all permissions are granted")
        } else {
            permissions.forEach {
                Log.i(TAG, "$it permission is missed")
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                requestPermissions(permissions.toTypedArray(), PERMISSION_REQUEST_CODE)
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        if (requestCode == PERMISSION_REQUEST_CODE) {
            for (i in permissions.indices) {
                Log.d(
                    TAG,
                    "${permissions[i]} is ${if (grantResults[i] != PackageManager.PERMISSION_GRANTED) "not " else ""}granted"
                )
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        return navController.navigateUp(appBarConfiguration) || super.onSupportNavigateUp()
    }

    companion object {

        fun showToast(message: CharSequence, context: Context?) {
            if (context != null) {
                val toast = Toast.makeText(context, message, Toast.LENGTH_SHORT)
                toast.show()
            }
        }

        fun initNewShell(): Shell {
            return ShellBuilder.sharedShell()
        }

        init {
            Shell.enableVerboseLogging = BuildConfig.DEBUG
        }
    }

    private fun onAutoMountChanged(isAutoMountRequired: Boolean) {
        if (isAutoMountRequired) {
            EasySSHFSService.start(applicationContext)
        } else {
            EasySSHFSService.stop(applicationContext)
        }
    }
}
